### PR TITLE
feature: add local disk quota

### PR DESCRIPTION
**Pull request by rudyfly**
### Ⅰ. Describe what this PR did
Add local disk quota. In alikernel, with dirquota feature to limit
directory block size.

### Ⅱ. Does this pull request fix one issue?

NONE

### Ⅳ. Describe how to verify it
In alikernel, you can use it like this:
set volume size is 10G
```bash
#pouch volume create -d local -n pouchvolume -o size=10g
CreatedAt:
Driver:       local
Mountpoint:
Name:         pouchvolume
Scope:
```
run a container, use df to see the volume size.
```bash
# pouch run -ti -v pouchvolume:/mnt --name test registry.hub.docker.com/library/busybox:latest df -h
Filesystem                Size      Used Available Use% Mounted on
overlay                  29.4G     19.6G      8.3G  70% /
tmpfs                    64.0M         0     64.0M   0% /dev
shm                      64.0M         0     64.0M   0% /dev/shm
tmpfs                    64.0M         0     64.0M   0% /run
/dev/sda3                10.0G      4.0K     10.0G   0% /mnt
```

### Ⅴ. Special notes for reviews

Signed-off-by: Rudy Zhang <rudyflyzhang@gmail.com>

### CODE REVIEW

File: volume/modules/local/quota.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+func GetDevID(dir string) (uint64, error) {
+	var st syscall.Stat_t
+	if err := syscall.Stat(dir, &st); err != nil {
+		logrus.Warnf("getDirDev: %s, %v", dir, err)
+		return 0, err
+	}
+	return st.Dev, nil
+}
+
+func toByteSize(s string) uint64 {
```
Comment (allencloud) [inline volume/modules/local/quota.go]: Please add a unit test for this function? @rudyfly 😄

File: volume/modules/local/quota.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	UseQuota     = true
+	quotaIDs     = make(map[uint32]uint32)
+	mountPoints  = make(map[uint64]string)
+	bytesPattern = regexp.MustCompile(`(?i)^(-?\d+)([KMGT]B?|B)$`)
+	gquota       = NewQuota()
+)
+
+// BaseQuota defines the quota operation interface.
+type BaseQuota interface {
+	QuotaDriverStart(dir string) (string, error)
```
Comment (allencloud) [inline volume/modules/local/quota.go]: From my own point of view, StartQuotaDriver is better than QuotaDriverStart. WDYT?

File: volume/modules/local/quota.go (hunk 1 above)
Comment (allencloud) [reply volume/modules/local/quota.go]: And this is similar to the function `ToBytes` in vendor package code.cloudfouncry.org/bytefmt/bytes.go.
We have a plan to remove this vendored package, could you encapsulate them into a package in pkg.
See https://github.com/alibaba/pouch/issues/356

File: volume/modules/local/quota.go (hunk 2 above)
Comment (rudyfly) [reply volume/modules/local/quota.go]: ok

File: volume/modules/local/grpquota.go [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+		return minID, err
+	}
+
+	lines := strings.Split(string(stdout), "\n")
+	for _, line := range lines {
+		if len(line) == 0 || line[0] != '#' {
+			continue
+		}
+		parts := strings.Split(line, " ")
+		if len(parts) == 0 {
```
Comment (allencloud) [inline volume/modules/local/grpquota.go]: I think 
```
if len(parts) == 0 {
    continue
}
```
is redundant.

File: volume/modules/local/quota.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+)
+
+const (
+	// QuotaMinID represents the minimize quota id.
+	QuotaMinID = uint32(16777216)
+)
+
+var (
+	lock        sync.Mutex
+	quotaLastID uint32
```
Comment (allencloud) [inline volume/modules/local/quota.go]: Could you please add more description or comments on these global variables in pkg local?
I think this is very useful for others to understand your thoughts. Actually currently I can read your code with logic, but have no idea about your thoughts.  😢

File: volume/modules/local/grpquota.go [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+
+	limit, err := bytefmt.ToKilobytes(size)
+	if err != nil {
+		return err
+	}
+
+	return quota.setUserQuota(id, limit, mountPoint)
+}
+
+// CheckMountpoint is used to check mount point.
```
Comment (allencloud) [inline volume/modules/local/grpquota.go]: Could you add one example valid line in /proc/mounts in comment please?
like:
```
// CheckMountpoint is used to check mount point which is in file /proc/mounts.
// one valid line must has the following format:
// cgroup /sys/fs/cgroup/hugetlb cgroup rw,nosuid,nodev,noexec,relatime,hugetlb 0 0
```
You can also tell others more about the exact meanings of each column. 😄

**Comment by allencloud** [comment]
That is a great work for you to complete this. @rudyfly 
Actually I am new for project quota, so I am still learning from that. I just left some tiny comments that may need your help.
And I wish to finish my learning and code review in the late today. Thanks a lot.

File: volume/modules/local/grpquota.go [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+		}
+		if err := quota.setUserQuota(0, 0, mountPoint); err != nil {
+			os.Remove(filename)
+			return mountPoint, err
+		}
+	}
+
+	// on
+	exit, stdout, stderr, err := exec.Run(0, "quotaon", "-pg", mountPoint)
+	if err != nil && exit != 1 {
```
Comment (allencloud) [inline volume/modules/local/grpquota.go]: Should this be `if err != nil || exit != 0`?
Since if the exit code is non-zero, it means running into an error.

**Comment by allencloud** [comment]
This is a great work. While I have one thing may need discussion.

> Is there any possibility that we make volume quota drive as plugins.

Currently we have grpquota as the default quota driver, and to be honest, I am afraid we have not taken full advantage of the interface `BaseQuota` we defined. In the code, I found we use a global instance of `gquota` and  function in pkg local `func StartQuotaDriver(dir string)` to transfer the request. We did not use `gquota.StartQuotaDriver(dir)` directly. And this way may weaken the usage of interface we defined and enhance the degree of code customization.

I am wondering if we could still 

* keep the interface `BaseQuota` 
* enhance the functionality of it to be capable of multiple specific quota driver plugins
* make users have ability to input the quota driver options

Maybe we can define a global variable `var QuotaDrivers map[string]BaseQuota`, and in every interface BaseQuota implementation(such as grpquota), we add an init function `init()` to register grpquota instance into QuotaDrivers. Then we can use another function like 
```
func GetQuotaDriver(dr string) (BaseQuota, error){
    if dr == "" {
        return grpquotaDriver,nil
    }
    driver, exist := QuotaDrivers(dr)
    if !exist{
        return fmt.Errorf("....")
    }
    return driver, nil
}
``` 
As a result, we could make users input their wanted quota drivers. And others could also add another quota driver via implementing the interface BaseQuota.

File: volume/modules/local/grpquota.go (hunk 3 above)
Comment (rudyfly) [reply volume/modules/local/grpquota.go]: yeah, it should be < 2

File: apis/server/volume_bridge.go [hunk 7]
```diff
@@ -41,12 +41,26 @@ func (s *Server) createVolume(ctx context.Context, rw http.ResponseWriter, req *
...
+
+	respVolume := types.VolumeInfo{
+		Name:       name,
+		Driver:     driver,
+		Labels:     config.Labels,
+		Mountpoint: volume.Path(),
+		CreatedAt:  volume.CreationTimestamp.Format("2006-1-2 15:04:05"),
+		Status:     make(map[string]interface{}),
+	}
+	for k, v := range volume.Options() {
```
Comment (allencloud) [inline apis/server/volume_bridge.go]: I think we can move the following codes above just before respVolume's initialization. We can use 
```
var status map[string]string
for k, v := range volume.Options() {
    if k != "" && v != "" {
        status[k] = v
    }
}
```

Then we can use this `status` in the `respVolume` initialization.

Although two ways can achieve the same effect, maybe we can make code more clear with two parts: data preparation and ``respVolume` initialization`. And this makes no separated assignments.

File: apis/server/volume_bridge.go [hunk 8]
```diff
@@ -61,10 +75,14 @@ func (s *Server) getVolume(ctx context.Context, rw http.ResponseWriter, req *htt
 		Mountpoint: volume.Path(),
 		CreatedAt:  volume.CreationTimestamp.Format("2006-1-2 15:04:05"),
 		Labels:     volume.Labels,
-		Status: map[string]interface{}{
-			"size": volume.Size(),
-		},
+		Status:     make(map[string]interface{}),
+	}
+	for k, v := range volume.Options() {
```
Comment (allencloud) [inline apis/server/volume_bridge.go]: The same as the previous comment.

File: apis/server/volume_bridge.go (hunk 7 above)
Comment (rudyfly) [reply apis/server/volume_bridge.go]: ok

File: volume/modules/local/grpquota.go (hunk 6 above)
Comment (rudyfly) [reply volume/modules/local/grpquota.go]: exit = 1 is also valid.

Comment (allencloud) [reply volume/modules/local/grpquota.go]: Oh, that is some kind of strange. Could you add some illustration here to tell code readers about that?

File: volume/modules/local/grpquota.go (hunk 5 above)
Comment (rudyfly) [reply volume/modules/local/grpquota.go]: ok

**Comment by allencloud** [comment]
LGTM, thanks again for the work. How about the test cases and disk quota document? @rudyfly

File: volume/modules/local/prjquota.go [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+	for _, line := range strings.Split(string(output), "\n") {
+		parts := strings.Split(line, " ")
+		if len(parts) != 6 {
+			continue
+		}
+		devID2, _ := GetDevID(parts[1])
+		if devID == devID2 {
+			mountPoint = parts[1]
+			fsType = parts[2]
+			for _, opt := range strings.Split(parts[3], ",") {
```
Comment (allencloud) [inline volume/modules/local/prjquota.go]: Could we just use 
```
if strings.Contains(parts[3], "prjquota") {
    hasQuota = true
}
```

File: volume/modules/local/grpquota.go [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+	for _, line := range strings.Split(string(output), "\n") {
+		parts := strings.Split(line, " ")
+		if len(parts) != 6 {
+			continue
+		}
+		devID2, _ := GetDevID(parts[1])
+		if devID == devID2 {
+			mountPoint = parts[1]
+			fsType = parts[2]
+			for _, opt := range strings.Split(parts[3], ",") {
```
Comment (allencloud) [inline volume/modules/local/grpquota.go]: Could we just use 
```
if strings.Contains(parts[3], "grpquota") {
    hasQuota = true
}
```
?

**Pull request merged**